### ISSUE TITLE
something about the turnover constraint

### ISSUE THREAD

**Issue by Bruce-yty** [jankrepl/deepdow#69]
These days I have been using your deepdow package to do some experiments about portfolio optimization. Thanks for your great work! 
But I have a problem here. I want to add the turnover rate constraint into the optimization. To achieve that, every round the network run over the training optimization, I have to keep the weights that has been calculated so that in the next round I can assure the new weight calculated won’t be too far from the previous one. 
So I want to ask you that if there is some way I can save the weight each time the network calculated during the training process?

**Comment by jankrepl** [comment]
Thank you for the question! It is a good one:)

### Answer 
One possible solution is to implement three things

- DataLoader that streams batches **without shuffling**. This way you know that if your original data was ordered in time then the batches will be just contiguous chunks of it. Note that both `RigidDataLoader` and `FlexibleDataLoader` shuffle.
- Write a `cvxpy` problem with one additional constraint representing the difference between the **current** and **previous** portfolio. Let's say `abs(w_curr - w_prev)` < threshold.
- In the forward pass we abandon batched allocation and simply **iterate sample by sample and solve the problem in each of these steps.**

**Minimal example**
So here we just take the variational form of `Softmax` and add to it the turnover constraint.
```python
import cvxpy as cp
from cvxpylayers.torch import CvxpyLayer
from deepdow.benchmarks import Benchmark
from deepdow.data import InRAMDataset
import numpy as np
import torch
from torch.utils.data import DataLoader


class Network(torch.nn.Module, Benchmark):
    def __init__(self, n_assets, threshold=0.2):
        super().__init__()

        x = cp.Parameter(n_assets)
        w_prev = cp.Parameter(n_assets)
        w_curr = cp.Variable(n_assets)
        obj = -x @ w_curr - cp.sum(cp.entr(w_curr))
        cons = [cp.sum(w_curr) == 1.,
                w_curr >= 0,
                cp.norm1(w_curr - w_prev) <= threshold,
                ]

        prob = cp.Problem(cp.Minimize(obj), cons)
        self.single_step_allocator = CvxpyLayer(prob, [x, w_prev], [w_curr])

    def forward(self, x, w_init=None):
        n_samples, n_channels, lookback, n_assets = x.shape
        device, dtype = x.device, x.dtype

        # some sofisticated feature extractor
        x = x.mean(dim=[1, 2])

        if w_init is None:
            w_prev = torch.ones(n_assets, device=device, dtype=dtype) / n_assets
        else:
            w_prev = w_init

        all_weights = []

        for i in range(n_samples):
            w_curr = self.single_step_allocator(x[i], w_prev)[0]
            all_weights.append(w_curr)
            w_prev = w_curr

        return torch.stack(all_weights, dim=0)


n_timesteps, n_assets = 400, 8
lookback, gap, horizon = 10, 0, 15
returns = np.random.random((n_timesteps, n_assets))
X_list, y_list = [], []

for i in range(lookback, n_timesteps - horizon - gap + 1):
    X_list.append(returns[i - lookback: i, :])
    y_list.append(returns[i + gap: i + gap + horizon, :])

X = np.stack(X_list, axis=0)[:, None, ...]
y = np.stack(y_list, axis=0)[:, None, ...]

batch_size = 10
dataset = InRAMDataset(X, y)
dataloader = DataLoader(dataset,
                        batch_size=batch_size,
                        shuffle=False  # The samples in a batch are contiguous in time
                        )

# Try it out for a single batch
dtype = torch.float32

network = Network(n_assets)
X_batch, _, _, _ = next(iter(dataloader))
w_init = torch.eye(n_assets, dtype=dtype)[2]  # we start with a portfolio invested solely in 3rd asset 

X_batch = X_batch.to(dtype)
network.to(dtype)

w = network(X_batch, w_init)
print(w)

```
Output (see the 3rd column)
```python
tensor([[0.0147, 0.0142, 0.9000, 0.0134, 0.0155, 0.0140, 0.0143, 0.0138],
        [0.0298, 0.0283, 0.8000, 0.0266, 0.0311, 0.0282, 0.0289, 0.0272],
        [0.0473, 0.0414, 0.7000, 0.0376, 0.0491, 0.0425, 0.0431, 0.0390],
        [0.0600, 0.0545, 0.6000, 0.0527, 0.0653, 0.0554, 0.0582, 0.0538],
        [0.0775, 0.0655, 0.5000, 0.0691, 0.0820, 0.0669, 0.0690, 0.0699],
        [0.0911, 0.0792, 0.4000, 0.0868, 0.0968, 0.0772, 0.0865, 0.0822],
        [0.1044, 0.0965, 0.3000, 0.1021, 0.1105, 0.0908, 0.0990, 0.0967],
        [0.1179, 0.1034, 0.2000, 0.1205, 0.1259, 0.0993, 0.1153, 0.1179],
        [0.1382, 0.1252, 0.1119, 0.1236, 0.1368, 0.1049, 0.1286, 0.1308],
        [0.1450, 0.1272, 0.1174, 0.1238, 0.1273, 0.1081, 0.1238, 0.1274]])

```

### Comments
Currently, all of the existing layers are designed to be independent across the sample dimension. In other words it does not matter how you shuffle samples in the batch, the layer never assumes the order is meaningful. However, as shown in the example, one can actually write a forward pass that assumes that the samples are contiguous in time.

If you observed some major turnover between two samples that are close in time maybe you can also try to implicitly enforce the similarity of predictions. There are more or less two ways. 

1. The first one is to use a (almost) *translation invariant* feature extractor like 1D convolution. Small shift of the rolling window should result in very similar extracted features.
2. The other technique is to make the horizon longer. Let's say the horizon is 30 (i.e days). If we move our sliding window by 1 day then the target **y** (future returns) slides by one day. Since the losses (i.e. `MeanReturns`) is looking at the entire horizon it means that (1+1)/30~6 % of the `future` has changed. There is a new entry from yesterday and also the 30th day from yesterday became 31st and is not inside of the horizon. This means the optimal allocation from yesterday should actually give a very similar loss today and therefore the network should not try to learn something fundamentally different.
